edgehog-device-runtime-docker/src/image.rs

```diff
@@ -0,0 +1,553 @@
+        Self {
+            name,
+            tag,
+            repo: Some(repo),
+        }
+    }
+
+    /// Get a reference to the image
+    pub fn tag(&self) -> String
+    where
+        S: Display,
+    {
+        self.to_string()
+    }
+
+    /// Pull the docker image struct.
+    ///
+    /// See the [Docker API reference](https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageCreate)
+    #[instrument]
+    pub async fn pull(&self, client: &Client) -> Result<(), ImageError>
+    where
+        S: Debug + Display,
+    {
+        let options = CreateImageOptions::from(self);
+
+        debug!("Creating the image {}", self);
+
+        client
+            // NOTE: Create the image by pulling it from the registry, here it's missing the root fs
+            //       option and the registry auth
+            .create_image(Some(options), None, None)
+            .try_for_each(|create_info| {
+                debug!("creating image: {:?}", create_info);
+
+                future::ready(Ok(()))
+            })
+            .await
+            .map_err(ImageError::Pull)?;
+
+        info!("Image {} created", self);
+
+        Ok(())
+    }
+
+    /// Inspect a docker image.
+    ///
+    /// See the [Docker API reference](https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageInspect)
+    #[instrument]
+    pub async fn inspect(&self, client: &Client) -> Result<Option<ImageInspect>, ImageError>
+    where
+        S: Debug + AsRef<str>,
+    {
+        let res = client.inspect_image(self.name.as_ref()).await;
+
+        debug!("received response {res:?}");
+
+        // Check if the image was not found
+        let inspect = match res {
+            Ok(inspect) => inspect,
+            Err(BollardError::DockerResponseServerError {
+                status_code: 404,
+                message,
+            }) => {
+                debug!("image not found: {message}");
+
+                return Ok(None);
+            }
+            Err(err) => return Err(ImageError::Inspect(err)),
+        };
+
+        debug!("inspected image: {inspect:?}");
+
+        Ok(Some(inspect))
+    }
+
+    /// Remove and image and it's parents.
+    ///
+    /// See the [Docker API reference](https://docs.docker.com/engine/api/v1.43/#tag/Image/operation/ImageDelete)
+    #[instrument]
+    pub async fn remove(
+        self,
+        client: &Client,
+    ) -> Result<Option<Vec<ImageDeleteResponseItem>>, ImageError>
+    where
+        S: Debug + Display,
+    {
+        let id = self.to_string();
+
+        debug!("removing image: {id}");
+
+        let res = client.remove_image(&id, None, None).await;
+
+        debug!(?res);
+
+        match res {
+            Ok(delete_res) => Ok(Some(delete_res)),
+            Err(BollardError::DockerResponseServerError {
+                status_code: 404,
+                message,
+            }) => {
+                warn!("image not found: {message}");
+
+                Ok(None)
```

Comment:
same

Reply:
Same as above, the image is not found so we don't need to remove it